File: 2025/11/main.py
#!/usr/bin/env python3

import os
import sys
from collections import deque

def findAllPaths(DATA):
    ans = 0
    queue = deque(['you'])
    while len(queue) > 0:
        cur = queue.pop()
        if cur == 'out':
            ans += 1
            continue
        newPaths = DATA[cur]
        for np in newPaths:
            queue.append(np)
    return ans

def doPart1():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    filename = f"{dir_path}/example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    DATA = {}
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            key, values = line.split(':')
            DATA[key.strip()] = [v.strip() for v in values.strip().split(' ')]
    ans = findAllPaths(DATA)
    print("Part 1:", ans)

def doPart2():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    filename = f"{dir_path}/example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue

def main():
    doPart1()
    doPart2()

if __name__ == "__main__":
    main()

File: 2025/11/main.cpp

#include "../../helpers/extras.hh"
#include <unordered_map>

std::unordered_map<std::string, std::vector<std::string>> DATA;
std::unordered_map<std::string, size_t> MEMORY;
size_t dfs(std::string cur, bool hasFFT, bool hasDAC)
{
    std::string key = cur;
    if (hasFFT) {
        key+="1";
    } else {
        key+="0";
    }
    if (hasDAC) {
        key+="1";
    } else {
        key+="0";
    }
    if (MEMORY.find(key) != MEMORY.end()) return MEMORY[key];

    if (cur == "out") {
        if (hasFFT && hasDAC) {
            return 1;
        } else {
            return 0;
        }
    }

    size_t curAns = 0;
    for (auto& np : DATA[cur]) {
        curAns += dfs(
            np,
            hasFFT || (np == "fft"),
            hasDAC || (np == "dac")
        );
    }
    MEMORY[key] = curAns;
    return curAns;
}

void doPart1(const char* filename)
{
    std::ifstream file(filename);
    std::string line;
    while (getline(file, line)) {
        stripString(line);
        if (line.length() == 0) continue;
        std::vector<std::string> sLine = splitString(line, ":");
        stripString(sLine[1]);
        for (auto val : splitString(sLine[1], " ")) {
            DATA[sLine[0]].push_back(val);
        }
    }
    file.close();

    MEMORY.clear();
    size_t ans = dfs("you", true, true);
    std::cout << "Part 1: " << ans << '\n';
}

void doPart2(const char* filename)
{
    (void)filename;
    MEMORY.clear();
    size_t ans = dfs("svr", false, false);
    std::cout << "Part 2: " << ans << '\n';
}

int main(int narg, char* args[])
{
    std::string filename = "input.dat";
    if (narg > 1) filename = std::string(args[1]);
    doPart1(filename.c_str());
    doPart2(filename.c_str());
    return 0;
}
